### DISCUSSION TITLE
Missionevents not using requiredDestinationTypes + modding compatibility

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#13941] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

For the longest time, transport missions could have unproper destinations, for example a VIP transport mission ending up in an empty location.

https://github.com/FakeFishGames/Barotrauma/issues/7497

This is caused by one thing:
**None of the missionevents check for adjacent location**

Missions have locations/destinations in 3 places:

- define locationtype in eventset `locationtype="outpost,city,research,military,mine"`, where this eventset triggers
- define requiredDestinationTypes in missionevent (NOT DEFINED) IF NOT VALID THE EVENT WILL NOT TRIGGER `requiredDestinationTypes="outpost,city,research,military,mine"`
- define potential destination in mission  under `<LocationType identifier="Outpost"/>`, will pick a destination where to place the mission

Example of a VIP transport mission between any of two outposts:

```xml

  <EventSet identifier="outpostmissionsexample" leveltype="outpost" locationtype="outpost,city,research,military,mine" >

    <ScriptedEvent identifier="missionevent_escort2coalition" commonness="80" requiredDestinationTypes="outpost,city,research,military,mine"`>
      <NPCWaitAction npctag="outpostmanager" wait="true" />
      <ConversationAction text="EventText.missionevent_escort2.c1" speakertag="outpostmanager" endeventifinterrupted="false" dialogtype="Small" />
      <MissionAction missiontag="escortVIPcoalition" />
      <NPCWaitAction npctag="outpostmanager" wait="false" />
    </ScriptedEvent>
```

# Second problem: no `requireddestinationfaction`. 

Cannot define a specific faction as destination. So for example some events implicitely state you are transporting passengers between two outposts of the same faction. Example: from military to military, same faction

`<eventtext.missionevent_escort3.c1>“We have a few prisoners that need to be transferred to a prison on another station. You'll be paid well for their transportation, but be aware that they need to be closely guarded.”</eventtext.missionevent_escort3.c1>`

# Third problem: mod compatibility (ANYOUTPOST)

For best compatibility, AnyOutpost should be defined as much as possible, but:

1. Anyoutpost is only used in missions. It cannot be used for defining event locations afaik https://github.com/search?q=repo%3AFakeFishGames%2FBarotrauma%20anyoutpost&type=code

2. You cannot rule out outpost locationtypes from Anyoupost. This means you cannot define outposts that you do NOT want to be picked up by anyoutpost (example: special "monster" outposts)

See
https://github.com/FakeFishGames/Barotrauma/issues/13482

------------------------

**Solution:** 

- Define `requiredDestinationTypes` on missionevents, ESPECIALLY TRANSPORT/CARGO MISSIONS
- Add `requireddestinationfaction`
- Add support for `Anyoupost` to eventsets/events
- Add support for `IsAnyOutpost="false"` in Locationtypes

**Side effects:**
When an eventset picks a missionevent, and this missionevent doesnt find valid destinations, it will not be triggered. This means in practice transport/cargo will become slightly less common, and less missions might be given (if eventset triggers 3 missionevents and 2 have invalid adjacent destinations you only have 1 triggering). Acceptable imo. 

### Version

v1.4.6.0 (Blood in the Water Update, hotfix 2)

**Comment by Regalis11** [comment]
Closing as a duplicate of https://github.com/FakeFishGames/Barotrauma/issues/7497

**Reply by Regalis11** [reply]
Addressed in our private repo now https://github.com/FakeFishGames/Barotrauma-development/pull/5341
